.github/workflows/csharp.yml

```diff
@@ -33,42 +51,37 @@ concurrency:
     group: C#-${{ github.head_ref || github.ref }}
     cancel-in-progress: true
 
+run-name:
+    # Set custom name if job is started manually and name is given
+    ${{ github.event_name == 'workflow_dispatch' && (inputs.name == '' && format('{0} @ {1}', github.ref_name, github.sha) || inputs.name) || '' }}
+
 jobs:
-    load-engine-matrix:
+    get-matrices:
         runs-on: ubuntu-latest
+        # Avoid running on schedule for forks
+        if: (github.repository_owner == 'valkey-io' || github.event_name != 'schedule') || github.event_name == 'push' || github.event_name == 'pull_request' || github.event_name == 'workflow_dispatch'
         outputs:
-          matrix: ${{ steps.load-engine-matrix.outputs.matrix }}
+            engine-matrix-output: ${{ steps.get-matrices.outputs.engine-matrix-output }}
+            host-matrix-output: ${{ steps.get-matrices.outputs.host-matrix-output }}
+            version-matrix-output: ${{ steps.get-matrices.outputs.version-matrix-output }}
         steps:
-            - name: Checkout
-              uses: actions/checkout@v4
-  
-            - name: Load the engine matrix
-              id: load-engine-matrix
-              shell: bash
-              run: echo "matrix=$(jq -c . < .github/json_matrices/engine-matrix.json)" >> $GITHUB_OUTPUT
-  
-    run-tests:
-        needs: load-engine-matrix
-        timeout-minutes: 25
+            - uses: actions/checkout@v4
+            - id: get-matrices
+              uses: ./.github/workflows/create-test-matrices
+              with:
+                  language-name: csharp
+                  # Run full test matrix if job started by cron or it was explictly specified by a person who triggered the workflow
+                  run-full-matrix: ${{ github.event.inputs.full-matrix || github.event_name == 'schedule' || 'false' }}
+
+    test-csharp:
+        needs: get-matrices
+        timeout-minutes: 35
```

Comment:
A side note — we should make a smart decision about the timeout we choose. 
First, extra time might make us wait longer for unresponsive test or workflow.
Second we need to have a trash hold that when crossed we need to do some thinking how to shorten and efficient the workflow so we won't end up with hours of waiting for each PR like in redis-rs.

Reply:
@barshaul @ikolomi 
Would appreciate your take on it.
Out of the scope of the PR.